9 backend new question

mogorovskiy opened the pull request.

mogorovskiy committed:
Add AnswerDto's and refactor QuesionGetById

mogorovskiy committed:
Refactor QuestionServiceImpl

mogorovskiy committed:
Add answer logic in QuestionServiceImpl.java

mogorovskiy committed:
mapToAnswerDto

mogorovskiy committed:
- add views and raiting fields in ANSWER table in DB
- add LocalDateTime to askedAt field in QuestionByIdDto
- modify imports

mogorovskiy committed:
- add one method in GlobalExceptionHandler
- imports optimized
- refactor QuestionServiceImpl

mogorovskiy committed:
- add method newQuestion in QuestionController
- add QuestionMapper
- add such method in QuestionServiceImpl and QuestionService
- add QuestionRequestDto

mogorovskiy committed:
- add method newQuestion in QuestionController
- add QuestionMapper
- add such method in QuestionServiceImpl and QuestionService
- add QuestionRequestDto
- rename createQuestion

mogorovskiy committed:
- add method newQuestion in QuestionController
- add QuestionMapper
- add such method in QuestionServiceImpl and QuestionService
- add QuestionRequestDto
- rename createQuestion
- delete answer's code

mogorovskiy committed:
- add createQuestion method (with setting author)
- optimize imports
- rename saveQuestion -> createQuestion method

mogorovskiy committed:
- add few setters
- now creating questions work correct (beta)

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/controller/QuestionController.java:
```diff
@@ -19,13 +23,18 @@ public class QuestionController {
         return questionService.getQuestionsPreviewPage(pageable);
     }
+
+    @PostMapping("/new")
```
Only using post method is clear that we will create something on the server, no need to add “new” on path
```suggestion
    @PostMapping
```

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/dto/answer/response/AnswerMainDto.java:
```diff
@@ -0,0 +1,15 @@
+@Data
+public class AnswerMainDto {
+
+    private long id;
```
```suggestion
    private Long id;
```

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/dto/question/response/QuestionByIdDto.java:
```diff
@@ -0,0 +1,20 @@
+@Data
+public class QuestionByIdDto {
+
+    private long id;
```
Same

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/service/QuestionService.java:
```diff
@@ -1,12 +1,16 @@
     QuestionPreviewPageResponseDto getQuestionsPreviewPage(Pageable pageable);
+
+    Question createQuestion(QuestionRequestDto questionRequestDto);
 }
```
Don’t return entity, return DTO instead

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/service/impl/QuestionServiceImpl.java:
```diff
@@ -1,36 +1,103 @@
         Optional<Question> questionOptional = questionRepository.findById(questionId);
-        return questionOptional.orElse(null);
+
+        return mapToQuestionDto(questionOptional.orElseThrow(() -> new NoSuchElementException("No question in id: " + questionId)));
```
What no question in id? Maybe “can’t find any question by given id:”?

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/model/Answer.java:
```diff
@@ -21,4 +22,10 @@ public class Answer extends BaseModel {
     private Question question;
+
+    @Column(nullable = false)
+    private Long views;
```
```suggestion
    private Long views = 0L;
```

koval-kn commented on backend/src/main/java/com/kk/heapoverflow/model/Answer.java:
```diff
@@ -21,4 +22,10 @@ public class Answer extends BaseModel {
+    private Long views;
+
+    @Column(nullable = false)
+    private Long rating;
```
```suggestion
    private Long rating = 0L;
```

mogorovskiy committed:
- deleted path on create question method
- correct id class
- QuestionService now returns DTO
- correct message when question was not found
- views and rating now is 0L by default
